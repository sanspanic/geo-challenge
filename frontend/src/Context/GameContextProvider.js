import React, { useState, useEffect } from "react";
import GameContext from "./GameContext";
import geoApi from "../API/geoAPI";

const GameContextProvider = ({ children }) => {
  const [speedBonus, setSpeedBonus] = useState(0);
  const [width, setWidth] = useState(100);
  const [level, setLevel] = useState(1);
  const [score, setScore] = useState(0);
  const [countries, setCountries] = useState([]);
  const [rank, setRank] = useState(1);
  const [hasLoaded, setHasLoaded] = useState(false);
  const [lives, setLives] = useState(5);
  const [finalScore, setFinalScore] = useState(0);

  const [status, setStatus] = useState({
    isActive: false,
    isLost: true,
    isWon: false,
  });

  useEffect(() => {
    const getCountryData = async () => {
      try {
        const res = await geoApi.getData();
        setCountries(res);
        setHasLoaded(true);
      } catch (err) {
        console.log(err);
      }
    };
    getCountryData();
  }, []);

  return (
    <GameContext.Provider
      value={{
        level,
        setLevel,
        score,
        setScore,
        countries,
        setCountries,
        hasLoaded,
        status,
        setStatus,
        lives,
        setLives,
        width,
        setWidth,
        speedBonus,
        setSpeedBonus,
        rank,
        setRank,
        finalScore,
        setFinalScore,
      }}
    >
      {children}
    </GameContext.Provider>
  );
};

export default GameContextProvider;
