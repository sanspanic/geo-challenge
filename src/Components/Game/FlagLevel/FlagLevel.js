import React, { useContext, useState, useEffect } from "react";
import GameContext from "../../../Context/GameContext";
import {
  makeUniqueSelection,
  pickWinner,
  isCorrect,
  gameWon,
  gameLost,
} from "../Common/helpers";
import Timer from "../Common/Timer";

const FlagLevel = () => {
  const { countries, hasLoaded, setScore, status, setStatus } = useContext(
    GameContext
  );
  const [selection, setSelection] = useState([
    { name: "", flag: "", capital: "" },
    { name: "", flag: "", capital: "" },
    { name: "", flag: "", capital: "" },
    { name: "", flag: "", capital: "" },
  ]);
  const [winner, setWinner] = useState({ name: "", flag: "", capital: "" });
  //each level will have 20 turns
  const [turn, setTurn] = useState(1);
  const [mistakes, setMistakes] = useState(["mistake"]);

  useEffect(() => {
    if (gameWon(turn)) {
      console.log("YA WON");
      setStatus({ ...status, isWon: true, isActive: false });
    }

    const indexArr = makeUniqueSelection();
    let selectionArr = [];
    for (let idx of indexArr) {
      selectionArr.push(countries[idx]);
    }
    if (hasLoaded) {
      setSelection(selectionArr);
      setWinner(pickWinner(selectionArr));
    }
  }, [hasLoaded, turn]);

  useEffect(() => {
    if (gameLost(mistakes)) {
      console.log("YA LOST");
      setStatus({ ...status, isLost: true, isActive: false });
    }
  }, [mistakes]);

  //1. get 4 random numbers between 0 and 249 - make sure these are unique
  //2. save selected four countries to state
  //3. pick random winner and save to state
  //4. render 4 flags

  const handleClick = (e) => {
    if (isCorrect(e.target.src, winner.flag)) {
      setScore((s) => s + 100);
    } else {
      setMistakes([...mistakes, "mistake"]);
      console.log(mistakes);
    }
    setTurn((turn) => turn + 1);
  };

  return (
    <>
      <h1 className="text-center font-black text-4xl col-start-2 col-span-6 mb-10">
        {winner.name}
      </h1>
      <div className="grid auto-rows-fr grid-cols-8 gap-2 place-items-stretch place-content-stretch">
        <img
          onClick={handleClick}
          alt="flag"
          src={selection[0].flag}
          className="shadow-xl rounded p-2 bg-gradient-gray mx-1 col-span-4 md:col-span-2  transform-gpu hover:scale-110 duration-500 cursor-pointer"
        />
        <img
          onClick={handleClick}
          alt="flag"
          src={selection[1].flag}
          className=" shadow-xl p-2 bg-gradient-gray mx-1 col-start-5 col-span-4 md:col-start-3 md:col-span-2  transform-gpu hover:scale-110 duration-500 cursor-pointer"
        />
        <img
          onClick={handleClick}
          alt="flag"
          src={selection[2].flag}
          className="shadow-xl p-2 bg-gradient-gray mx-1 col-start-1 col-span-4 md:col-start-5 md:col-span-2  transform-gpu hover:scale-110 duration-500 cursor-pointer"
        />
        <img
          onClick={handleClick}
          alt="flag"
          src={selection[3].flag}
          className="shadow-xl p-2 bg-gradient-gray mx-1 col-start-5 col-span-4 md:col-start-7 md:col-span-2  transform-gpu hover:scale-110 duration-500 cursor-pointer"
        />
      </div>
      <Timer mistakes={mistakes} />
    </>
  );
};

export default FlagLevel;
