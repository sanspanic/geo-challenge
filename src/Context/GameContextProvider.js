import React, { useState } from "react";
import GameContext from "./GameContext";

const GameContextProvider = ({ children }) => {
  const [level, setLevel] = useState({ num: 1, title: "Guess The Flag" });
  const [score, setScore] = useState(0);

  return (
    <GameContext.Provider value={{ level, setLevel, score, setScore }}>
      {children}
    </GameContext.Provider>
  );
};

export default GameContextProvider;
