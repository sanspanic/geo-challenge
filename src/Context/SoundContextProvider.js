import React, { useState } from "react";
import SoundContext from "./SoundContext";
import useSound from "use-sound";
import correct from "../Assets/sounds/correct.wav";
import wrong from "../Assets/sounds/wrong.wav";
import levelup from "../Assets/sounds/levelup.wav";
import gamelost from "../Assets/sounds/gamelost.wav";
import gamewon from "../Assets/sounds/gamewon.wav";

const SoundContextProvider = ({ children }) => {
  const [playCorrect] = useSound(correct, { volume: 0.25 });
  const [playWrong] = useSound(wrong, { volume: 0.25 });
  const [playLevelUp] = useSound(levelup, { volume: 0.25 });
  const [playGameOver] = useSound(gamelost, { volume: 0.25 });
  const [playGameWon] = useSound(gamewon, { volume: 0.55 });

  return (
    <SoundContext.Provider
      value={{ playCorrect, playWrong, playLevelUp, playGameOver, playGameWon }}
    >
      {children}
    </SoundContext.Provider>
  );
};

export default SoundContextProvider;
