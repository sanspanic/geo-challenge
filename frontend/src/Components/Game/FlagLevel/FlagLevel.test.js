import React from "react";
import { render } from "@testing-library/react";
import GameContextProvider from "../../../Context/GameContextProvider";
import SoundContextProvider from "../../../Context/SoundContextProvider";
import FlagLevel from "./FlagLevel";

test("renders without crashing", () => {
  render(
    <GameContextProvider>
      <SoundContextProvider>
        <FlagLevel />
      </SoundContextProvider>
    </GameContextProvider>
  );
});

test("matches snapshot", () => {
  const { asFragment } = render(
    <GameContextProvider>
      <SoundContextProvider>
        <FlagLevel />
      </SoundContextProvider>
    </GameContextProvider>
  );
  expect(asFragment()).toMatchSnapshot();
});
