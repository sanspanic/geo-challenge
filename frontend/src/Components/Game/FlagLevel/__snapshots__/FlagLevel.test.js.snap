// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <h1
    class="text-center font-black text-4xl col-start-2 col-span-6 mb-10"
  />
  <div
    class="grid flags-grid grid-cols-8 gap-2 place-content-stretch place-items-stretch"
  >
    <img
      alt="flag"
      class=" w-40 mx-1 col-span-4 md:col-span-2 transform-gpu hover:scale-110 duration-500 cursor-pointer"
      src=""
    />
    <img
      alt="flag"
      class=" w-40 mx-1 col-start-5 col-span-4 md:col-start-3 md:col-span-2 transform-gpu hover:scale-110 duration-500 cursor-pointer"
      src=""
    />
    <img
      alt="flag"
      class=" w-40 mx-1 col-start-1 col-span-4 md:col-start-5 md:col-span-2  transform-gpu hover:scale-110 duration-500 cursor-pointer"
      src=""
    />
    <img
      alt="flag"
      class=" w-40 mx-1 col-start-5 col-span-4 md:col-start-7 md:col-span-2  transform-gpu hover:scale-110 duration-500 cursor-pointer"
      src=""
    />
  </div>
  <div
    class="w-40 my-6 border-2 border-cerise-500 rounded-xl bg-gradient-pale p-2 bg-gradient-gray"
  >
    <div
      class="text-center"
    >
      Score: 
      <span
        class="text-xl font-bold"
      >
        0
      </span>
    </div>
    <div
      class="text-center"
    >
      <svg
        class="inline text-cerise-500 my-1"
        fill="currentColor"
        height="24"
        viewBox="0 0 256 256"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <path
          d="M133.65683,211.88227l81.0323-81.0322c19.9104-19.91038,22.84784-52.666,4.00583-73.59029a52.0026,52.0026,0,0,0-75.46451-2.02934L127.99994,70.46082,114.85007,57.3109C94.93972,37.40047,62.18409,34.463,41.25979,53.305a52.00261,52.00261,0,0,0-2.02934,75.46452l83.11268,83.11272A8,8,0,0,0,133.65683,211.88227Z"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
      </svg>
      <svg
        class="inline text-cerise-500 my-1"
        fill="currentColor"
        height="24"
        viewBox="0 0 256 256"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <path
          d="M133.65683,211.88227l81.0323-81.0322c19.9104-19.91038,22.84784-52.666,4.00583-73.59029a52.0026,52.0026,0,0,0-75.46451-2.02934L127.99994,70.46082,114.85007,57.3109C94.93972,37.40047,62.18409,34.463,41.25979,53.305a52.00261,52.00261,0,0,0-2.02934,75.46452l83.11268,83.11272A8,8,0,0,0,133.65683,211.88227Z"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
      </svg>
      <svg
        class="inline text-cerise-500 my-1"
        fill="currentColor"
        height="24"
        viewBox="0 0 256 256"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <path
          d="M133.65683,211.88227l81.0323-81.0322c19.9104-19.91038,22.84784-52.666,4.00583-73.59029a52.0026,52.0026,0,0,0-75.46451-2.02934L127.99994,70.46082,114.85007,57.3109C94.93972,37.40047,62.18409,34.463,41.25979,53.305a52.00261,52.00261,0,0,0-2.02934,75.46452l83.11268,83.11272A8,8,0,0,0,133.65683,211.88227Z"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
      </svg>
      <svg
        class="inline text-cerise-500 my-1"
        fill="currentColor"
        height="24"
        viewBox="0 0 256 256"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <path
          d="M133.65683,211.88227l81.0323-81.0322c19.9104-19.91038,22.84784-52.666,4.00583-73.59029a52.0026,52.0026,0,0,0-75.46451-2.02934L127.99994,70.46082,114.85007,57.3109C94.93972,37.40047,62.18409,34.463,41.25979,53.305a52.00261,52.00261,0,0,0-2.02934,75.46452l83.11268,83.11272A8,8,0,0,0,133.65683,211.88227Z"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
      </svg>
      <svg
        class="inline text-cerise-500 my-1"
        fill="currentColor"
        height="24"
        viewBox="0 0 256 256"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <path
          d="M133.65683,211.88227l81.0323-81.0322c19.9104-19.91038,22.84784-52.666,4.00583-73.59029a52.0026,52.0026,0,0,0-75.46451-2.02934L127.99994,70.46082,114.85007,57.3109C94.93972,37.40047,62.18409,34.463,41.25979,53.305a52.00261,52.00261,0,0,0-2.02934,75.46452l83.11268,83.11272A8,8,0,0,0,133.65683,211.88227Z"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
      </svg>
    </div>
  </div>
  <div
    class="self-stretch col-start-1 col-span-8 border-2 border-black"
  >
    <div
      class="bg-gradient-purple timer"
      style="width: 100%;"
    />
  </div>
</DocumentFragment>
`;
